File: Easy/0257_binary-tree-paths/binary-tree-paths.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func binaryTreePaths(root *TreeNode) []string {
  var paths []string

	var dfs func(node *TreeNode, path string)
	dfs = func(node *TreeNode, path string) {
		if node == nil {
			return
		}

		if path == "" {
			path = fmt.Sprintf("%d", node.Val)
		} else {
			path = fmt.Sprintf("%s->%d", path, node.Val)
		}

		if node.Left == nil && node.Right == nil {
			paths = append(paths, path)
			return
		}

		dfs(node.Left, path)
		dfs(node.Right, path)
	}
	dfs(root, "")
	return paths
}

func BuildTree(data string) *TreeNode {
	if data == "" {
		return nil
	}

	values := strings.Split(data, ",")
	if len(values) == 0 || values[0] == "null" {
		return nil
	}

	val, err := strconv.Atoi(values[0])
	if err != nil {
		return nil
	}
	root := &TreeNode{Val: val}

	queue := []*TreeNode{root}
	i := 1

	for len(queue) > 0 && i < len(values) {
		node := queue[0]
		queue = queue[1:]

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Left = &TreeNode{Val: v}
				queue = append(queue, node.Left)
			}
		}
		i++

		if i < len(values) && values[i] != "null" {
			if v, err := strconv.Atoi(values[i]); err == nil {
				node.Right = &TreeNode{Val: v}
				queue = append(queue, node.Right)
			}
		}
		i++
	}
	return root
}

func printTree(node *TreeNode) {
	if node == nil {
		return
	}
	printTree(node.Left)
	fmt.Printf("%d -> ", node.Val)
	printTree(node.Right)
}

func main() {
	tests := []string{
		"4,2,7,1,3,6,9",
		"2,1,3",
		"",
	}

	for _, t := range tests {
		tree := BuildTree(t)
		fmt.Printf("---\ntree: ")
		printTree(tree)
		fmt.Printf("\n binaryTreePaths: %s\n", binaryTreePaths(tree))
	}
}

File: Easy/0257_binary-tree-paths/binary-tree-paths.py
#!/bin/python3

from typing import List, Optional
from dataclasses import dataclass
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        paths = []

        def dfs(node, path):
            if not node:
                return
            path += str(node.val)
            if not node.left and not node.right:
                paths.append(path)
            else:
                path += "->"
                dfs(node.left, path)
                dfs(node.right, path)
        dfs(root, "")
        return paths

def buildTree(data: str) -> TreeNode:
    if not data:
        return None
    values = data.split(",")
    if not values or values[0] == "null":
        return None
    root = TreeNode(int(values[0]))
    queue = deque([root])
    i = 1
    while queue and i < len(values):
        node = queue.popleft()
        if i < len(values) and values[i] != "null":
            node.left = TreeNode(int(values[i]))
            queue.append(node.left)
        i += 1
        if i < len(values) and values[i] != "null":
            node.right = TreeNode(int(values[i]))
            queue.append(node.right)
        i += 1
    return root

def printTree(node: TreeNode) -> None:
    if not node:
        return
    printTree(node.left)
    print(node.val,end=' ')
    printTree(node.right)

def main():
    tests: List[str] = [
        "1,2,3,null,5",
        "1",
    ]

    for t in tests:
        tree = buildTree(t)
        print(f"---\ntree: ",end='')
        printTree(tree)
        print(f"\nbinaryTreePaths: {Solution().binaryTreePaths(tree)}")
 
if __name__ == "__main__":
    main()

